README.md
<h1 align="center">elaphureLink</h1>

<div align="center">

[![Build status][github-action-build-image]][github-action-build-url] [![license][license-image]][license-url]

[github-action-build-image]: https://github.com/windowsair/elaphureLink/actions/workflows/build-binary.yml/badge.svg
[github-action-build-url]: https://github.com/windowsair/elaphureLink/actions/workflows/build-binary.yml

[license-image]: https://img.shields.io/badge/license-BSD2-green.svg
[license-url]: https://github.com/windowsair/elaphureLink/blob/master/LICENSE.md

elaphureLink is a CMSIS-DAP compliant debugger solution for Arm® Keil® µVision® IDE that provides conversion between TCP/IP and DAP.


</div>


## Features

- Easy to use.
- Support for custom debug hardware implementation.
- Debug port support
    - [x] SWD
    - [ ] JTAG(Coming Soon)
- Debug spec support
  - [x] ADIv5

## Quick Start

This video shows how to install the driver, connect to the DAP device and perform the flash program.


https://user-images.githubusercontent.com/17078589/183233113-5dcac5f6-47d6-4d45-a231-12698253b3e2.mp4


## System Requirements

- Supported OS: Windows 7 or later.
- Supported Platforms: x86-64(AMD64).
- Additional requirements
  - [Microsoft Visual C++ 2015-2019 Redistributable x86 and x64](https://docs.microsoft.com/en-us/cpp/windows/redistributing-visual-cpp-files?view=msvc-170)
  - [.Net Framework 4.7.2](https://dotnet.microsoft.com/en-us/download/dotnet-framework/net472) (For elaphureLink.Wpf).
- µVision tested version
  - [x] v5.12.0.0
  - [x] v5.16a
  - [x] v5.25.2.0
  - [x] v5.31.0.0
  - [x] v5.37.0.0

> Due to the large number of versions, only some versions of µVision were tested.

## Download and Installation

The required dependencies are already packaged in the elaphureLink.Wpf project to provide the end user with an out-of-the-box experience.

**Stable Version**
Stable and pre-release versions are available in [Releases](https://github.com/windowsair/elaphureLink/releases).


**Develop Version**
You can download the latest build in Github Action at [build-binary](https://github.com/windowsair/elaphureLink/actions/workflows/build-binary.yml).


**Update**
You can check and update the latest stable version in the Info Page of elaphureLink.Wpf.

> At present, only the distribution of portable versions is available for download. All you need to do is extract and run `elaphureLink.Wpf.exe`


## Application case

- [wireless-esp8266-dap](https://github.com/windowsair/wireless-esp8266-dap) Single board wireless debugger based on ESP8266.

> Do you want to add a new case? Start a new pull request or issue!

----

## Develop

This section will introduce the relevant content in development.

### Build

Visual Studio 2019 or later version is recommended for building. Currently, you can only build the `elaphure.Wpf` project, and other dependencies will be built automatically.


### Architecture

The following diagram shows the system architecture of elaphureLink.

![Architecture](docs/elaphureLink_architecture.png)

elaphureLink consists of the following 5 main components:

- `AGDI` elaphureLinkAGDI is application generic debug interface of elaphureLink. It is the bridge between Arm Keil µVision and RDDI connection. It performs the interface to all basic debugger features, allows complex breakpoints, flash programming or access to system memory. Some of the debug settings and GUI in µVision are also provided by it.

- `RDDI` elaphureLinkRDDI is remote device debug interface of elaphureLink that allow remote DAP device communicate with the debug interface. It converts the requests into standard CMSIS-DAP commands and communicates with the proxy.

- `Proxy` elaphureLinkProxy play the role of communicating with the debug unit. It exports some API for connection management. Developers can use these API to build specific applications.

- `Wpf` elaphureLink.Wpf is a simple GUI wrapper of elaphureLinkProxy. It provides a number of utilities for managing drivers and connection control.

- `Server Provides` It may be the various devices containing the CMSIS-DAP implementation to provide debugging capabilities. For more examples, see [Application case](#Application-case).


### Proxy and device develop

During each connection, a specific protocol needs to be followed between elaphureLinkProxy and debug unit. For more information about the protocol, see [proxy_protocol](docs/proxy_protocol.md).


You may want to build your own application using elaphureLinkProxy, in this case please refer to the Proxy API, available at [Proxy API](docs/proxy_api.md).


## Credit

[@Mouri_Naruto](https://github.com/MouriNaruto) He provided unique insights into Windows programming for this project.


## License

[elaphureLink License](License.md)


elaphureLinkAGDI/ChangeLog.txt
/**************************************************************************//**
 *
 *   DbgCM.DLL Code Template: Change Log
 *
 *   Date : $Date: 2020-09-02 09:57:33 +0200 (Wed, 02 Sep 2020) $
 *
 *   Note : Latest revisions on top
 *
 ******************************************************************************/


Version 1.3.4 (02/09/2020)
==========================
- Enhancements:
   - Added System Description File (SDF) support.
   - Added basic CoreSight Cross Trigger Interface (CTI) support.
   - Added CoreSight Trace Funnel support.
   - Added Embedded Trace Buffer (ETB and TMC in ETB mode) support for
     ITM/DWT trace and Trace Records window.
   - Added LinkCom() function to easier demonstrate usage of debug unit
     communication lock. For example to allow safe access to a different
     AP than used for CPU debug.

- Bug fixes:
   - Fixed incorrect error generation in AG_DapAcc(), AGDI.cpp.
   - Fixed selection of TPIU formatter for some CPUs.
   - Fixed missing sequence call for DebugCoreStop (always executed
     hard-coded default).
   - Fixed order of Trace Uninit and Core Debug Stop. Trace Uninit must
     happen first in case system design turns off trace domain clocks when
     clearing DEMCR.TRCENA.
   - Removed ROM Table error for unknown components in CPU PN based
     detection scheme.



Version 1.3.3 (28/01/2020)
==========================
- Enhancements:
   - Added support for AG_CB_DEVICECHECK to check the connected device.
   - Support for "dormant" Bit in "__protocol" debug access variable.
     (CMSIS Pack Debug Support).
   - Changes to add tracepoint support.

- Bug fixes:
   - Fixed progress bar for flash download handling which made uVision
     sometimes appear to be frozen although it was still downloading.
   - Fixed crash in SBC command when used in INI script for flash download.



Version 1.3.2 (02/07/2019)
==========================
- MEM-AP support:
   - AP_CONTEXTs to handle different AP types in a debug connection.
   - Deprecated use of 'CSW_Val' and 'CSW_Val_Base'. Use AP_CONTEXT::CSW_Val_Base
     instead (see "Usage of AP_Context" in JTAG.CPP/SWD.CPP).
   - Added AXI-AP support.

- AG_MemAccX:
   - Refined AG_MemAccX() and supporting functions for better maintainability.
   - Removed automatic update of AP_Sel when writing to DP SELECT in AG_MemAccX().
   - Added ACCMX_ANYSZ to enable explicit secure / non-secure accesses without
     fixed access size.

- IP Support:
   - Added Cortex-M35P support.
   - Don't use DP CTRL/STAT CDBGRSTREQ bit. Only a last resort to recover locked devices.

- Enhancements:
   - Source code comments on "Usage of AP_Context" and core register access functions.
   - Extended target debug key fallback with newer ULINK AGDI DLLs.
   - Extend memory access error messages with more details (SetStatusMem()).
   - Added reset type and connection type in "__connection" debug access variable.
   - Aligned PDSCDebug_ResetCatchClear() with corresponding CMSIS default sequence.
   - Trace clock setting separate from CPU clock setting.
   - Skip check for trace hardware availability during flash download.
   - Trace errors are printed red in status bar.
   
- Bug Fixes:
   - Fixed v8-M core register accesses.
   - Fixed crashes in debug access recovery feature after connection loss.
   - Fixed SetClrSwBP() return code check in SaCBreaks().
   - Fixed uVision window updated calls.
   - Fixed Trace Setup Dialog around ITM Privileged edit field.



Version 1.3.1 (09/01/2018)
==========================
- Enhanced connection flow for Flash Download "without stop".
   - No implicit reset on Flash Download if without stop.
   - Actively stop CPU if Flash Download without stop.
   - Disable all exceptions (incl. Hard Fault) if Flash Download without stop.
- Deactivated demo data code in TrcExc.cpp by default.

- Bug Fixes:
  - Fixed hard shutdown scenario.
  - v8-M Trace: DWT_CTRL.CYCDISS Bit unexpectedly modified during DWT_CTRL access.
  - Fixed incorrectly forwarded buffer length in VerifyMem().



Version 1.3.0 (17/11/2016)
==========================
- Added optional support for the following features (see defines in COLLECT.H):
  - Debug and Trace for v8-M based CPUs (Cortex-M23 and Cortex-M33).
    - Architecture support.
    - Relax connection aborts if debugger has no access to secure resources.
  - CMSIS Pack Debug Description (PDSCDebug module).
  - Extended AGDI target access functions (AG_MemAccX(), AG_DapAcc() functions).
  - Device State Monitor, detects changes of CPU state while running and stopped.
  - Connect "without Stop".
  - Debug Access Recovery.
  
- Enhancements and Bug Fixes:
  - Fixes to breakpoint handling (functionality and stability). Also addresses
    incorrect line stepping behaviour.
  - Skip BKPT instruction at start of go/step operation.
  - Explicit initialization of global variables.
  - "Stop After Reset"-option (permanent reset vector catch).
  - "Any" Debug Unit selection, selects first detected Debug Unit.
  - Fixed potential race condition between target stop and AGDI DLL memory cache.
  - Support for new CoreSight DAPs (JTAG IDCODEs and SW DPIDR values).
  - Debug (Breakpoint) Resource Manager, immediately checks availability of
    Breakpoints and Tracepoints.
  - SwBreakConfig (SBC) command.
  - UVSock updates for ITM trace.
  - Missing Trace Exception window reset between debug sessions.
  - Copy default options from ULINK2 settings if available.
  - Error message box enhancements.

Version 1.2.8 (29/09/2015)
==========================
- Cortex-M7 Support
  - Fixed M7 r0p0/r0p1 MASKINTS workaround artificially causing HALTED bit in DSFR
    after breakpoint hit.
  - Architecture-compliant programming of DWT (DWT_FUNCTION.FUNC of linked address
    comparator must be 0). Fixes linked Cortex-M7 access breakpoints ("variable == value").

src/DbgCM/ELF.h
/**************************************************************************/ /**
 *           Cortex-M Middle/Upper layer Debug driver Template for µVision
 *
 * @version  V1.0.1
 * @date     $Date: 2015-04-28 15:09:20 +0200 (Tue, 28 Apr 2015) $
 *
 * @note
 * Copyright (C) 2009-2015 ARM Limited. All rights reserved.
 *
 * @par
 * ARM Limited (ARM) is supplying this software for use with Keil uVision
 * and Cortex-M processor based microcontrollers.
 *
 * @par
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL COPYRIGHT HOLDERS AND CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 ******************************************************************************/



#ifndef __ELF_H__
#define __ELF_H__


#define ELF32_ST_BIND(i)    ((i) >> 4)
#define ELF32_ST_TYPE(i)    ((i)&0x0F)
#define ELF32_ST_INFO(b, t) (((b) << 4) + ((t)&0x0F))


/* Fields in e_ident[] */

#define EI_MAG0      0    /* File identification byte 0 index */
#define ELFMAG0      0x7F /* Magic number byte 0 */

#define EI_MAG1      1   /* File identification byte 1 index */
#define ELFMAG1      'E' /* Magic number byte 1 */

#define EI_MAG2      2   /* File identification byte 2 index */
#define ELFMAG2      'L' /* Magic number byte 2 */

#define EI_MAG3      3   /* File identification byte 3 index */
#define ELFMAG3      'F' /* Magic number byte 3 */

#define EI_CLASS     4 /* File class */
#define ELFCLASSNONE 0 /* Invalid class */
#define ELFCLASS32   1 /* 32-bit objects */
#define ELFCLASS64   2 /* 64-bit objects */

#define EI_DATA      5 /* Data encoding */
#define ELFDATANONE  0 /* Invalid data encoding */
#define ELFDATA2LSB  1 /* 2's complement, little endian */
#define ELFDATA2MSB  2 /* 2's complement, big endian */

#define EI_VERSION   6 /* File version */

//---TODO: format

#define EI_OSABI            7   /* Operating System/ABI indication */
#define ELFOSABI_NONE       0   /* UNIX System V ABI */
#define ELFOSABI_HPUX       1   /* HP-UX operating system */
#define ELFOSABI_NETBSD     2   /* NetBSD */
#define ELFOSABI_LINUX      3   /* GNU/Linux */
#define ELFOSABI_HURD       4   /* GNU/Hurd */
#define ELFOSABI_SOLARIS    6   /* Solaris */
#define ELFOSABI_MONTEREY   7   /* Monterey */
#define ELFOSABI_IRIX       8   /* IRIX */
#define ELFOSABI_FREEBSD    9   /* FreeBSD */
#define ELFOSABI_TRU64      10  /* TRU64 UNIX */
#define ELFOSABI_MODESTO    11  /* Novell Modesto */
#define ELFOSABI_OPENBSD    12  /* OpenBSD */
#define ELFOSABI_STANDALONE 255 /* Standalone (embedded) application */
#define ELFOSABI_ARM        97  /* ARM */

#define EI_ABIVERSION       8 /* ABI version */
#define EI_PAD              9 /* Start of padding bytes */


/* Values for e_type, which identifies the object file type */

#define ET_NONE   0x0000 /* No file type */
#define ET_REL    0x0001 /* Relocatable file */
#define ET_EXEC   0x0002 /* Executable file */
#define ET_DYN    0x0003 /* Shared object file */
#define ET_CORE   0x0004 /* Core file */
#define ET_LOOS   0xFE00 /* Operating system-specific */
#define ET_HIOS   0xFEFF /* Operating system-specific */
#define ET_LOPROC 0xFF00 /* Processor-specific */
#define ET_HIPROC 0xFFFF /* Processor-specific */


/* Values for e_machine, which identifies the architecture */

#define EM_NONE        0                                                   /* No machine */
#define EM_M32         1                                                   /* AT&T WE 32100 */
#define EM_SPARC       2                                                   /* SUN SPARC */
#define EM_386         3                                                   /* Intel 80386 */
#define EM_68K         4                                                   /* Motorola m68k family */
#define EM_88K         5                                                   /* Motorola m88k family */
#define EM_486         6                                                   /* Intel 80486 */
#define EM_860         7                                                   /* Intel 80860 */
#define EM_MIPS        8                                                   /* MIPS R3000 (officially, big-endian only) */
#define EM_S370        9                                                   /* IBM System/370 */
#define EM_MIPS_RS4_BE 10 /* MIPS R4000 big-endian */                      /* Depreciated */
#define EM_MIPS_RS3_LE 10 /* MIPS R3000 little-endian (Oct 4 1999 Draft)*/ /* Depreciated */

#define EM_PARISC      15 /* HPPA */
#define EM_VPP550      17 /* Fujitsu VPP500 */
#define EM_SPARC32PLUS 18 /* Sun's "v8plus" */
#define EM_960         19 /* Intel 80960 */
#define EM_PPC         20 /* PowerPC */
#define EM_PPC64       21 /* 64-bit PowerPC */

#define EM_V800        36                       /* NEC V800 series */
#define EM_FR20        37                       /* Fujitsu FR20 */
#define EM_RH32        38                       /* TRW RH32 */
#define EM_MCORE       39 /* Motorola M*Core */ /* May also be taken by Fujitsu MMA */
#define EM_RCE         39                       /* Old name for MCore */
#define EM_ARM         40                       /* ARM */
#define EM_OLD_ALPHA   41                       /* Digital Alpha */
#define EM_SH          42                       /* Hitachi SH */
#define EM_SPARCV9     43                       /* SPARC v9 64-bit */
#define EM_TRICORE     44                       /* Siemens Tricore embedded processor */
#define EM_ARC         45                       /* Argonaut RISC Core, Argonaut Technologies Inc. */
#define EM_H8_300      46                       /* Hitachi H8/300 */
#define EM_H8_300H     47                       /* Hitachi H8/300H */
#define EM_H8S         48                       /* Hitachi H8S */
#define EM_H8_500      49                       /* Hitachi H8/500 */
#define EM_IA_64       50                       /* Intel IA-64 Processor */
#define EM_MIPS_X      51                       /* Stanford MIPS-X */
#define EM_COLDFIRE    52                       /* Motorola Coldfire */
#define EM_68HC12      53                       /* Motorola M68HC12 */
#define EM_MMA         54                       /* Fujitsu Multimedia Accelerator */
#define EM_PCP         55                       /* Siemens PCP */
#define EM_NCPU        56                       /* Sony nCPU embedded RISC processor */
#define EM_NDR1        57                       /* Denso NDR1 microprocesspr */
#define EM_STARCORE    58                       /* Motorola Star*Core processor */
#define EM_ME16        59                       /* Toyota ME16 processor */
#define EM_ST100       60                       /* STMicroelectronics ST100 processor */
#define EM_TINYJ       61                       /* Advanced Logic Corp. TinyJ embedded processor */

#define EM_FX66        66 /* Siemens FX66 microcontroller */
#define EM_ST9PLUS     67 /* STMicroelectronics ST9+ 8/16 bit microcontroller */
#define EM_ST7         68 /* STMicroelectronics ST7 8-bit microcontroller */
#define EM_68HC16      69 /* Motorola MC68HC16 Microcontroller */
#define EM_68HC11      70 /* Motorola MC68HC11 Microcontroller */
#define EM_68HC08      71 /* Motorola MC68HC08 Microcontroller */
#define EM_68HC05      72 /* Motorola MC68HC05 Microcontroller */
#define EM_SVX         73 /* Silicon Graphics SVx */
#define EM_ST19        74 /* STMicroelectronics ST19 8-bit microcontroller */
#define EM_VAX         75 /* Digital VAX */

#define EM_PJ          99 /* picoJava */


/* Values for e_version */

#define EV_NONE    0 /* Invalid ELF version */
#define EV_CURRENT 1 /* Current version */


/* Values for section header, sh_type field */
#define SHT_NULL          0  /* Section header table entry unused */
#define SHT_PROGBITS      1  /* Program specific (private) data */
#define SHT_SYMTAB        2  /* Link editing symbol table */
#define SHT_STRTAB        3  /* A string table */
#define SHT_RELA          4  /* Relocation entries with addends */
#define SHT_HASH          5  /* A symbol hash table */
#define SHT_DYNAMIC       6  /* Information for dynamic linking */
#define SHT_NOTE          7  /* Information that marks file */
#define SHT_NOBITS        8  /* Section occupies no space in file */
#define SHT_REL           9  /* Relocation entries, no addends */
#define SHT_SHLIB         10 /* Reserved, unspecified semantics */
#define SHT_DYNSYM        11 /* Dynamic linking symbol table */

#define SHT_INIT_ARRAY    14 /* Array of pointers to init functions */
#define SHT_FINI_ARRAY    15 /* Array of pointers to finish functions */
#define SHT_PREINIT_ARRAY 16 /* Array of pointers to pre-init functions */

#define SHT_LOOS          0x60000000 /* Operating system specific semantics, lo */
#define SHT_HIOS          0x6fffffff /* Operating system specific semantics, hi */


/* The next three section types are defined by Solaris, and are named
   SHT_SUNW*.  We use them in GNU code, so we also define SHT_GNU*
   versions.  */

#define SHT_SUNW_verdef  0x6ffffffd /* Versions defined by file */
#define SHT_SUNW_verneed 0x6ffffffe /* Versions needed by file */
#define SHT_SUNW_versym  0x6fffffff /* Symbol versions */

#define SHT_GNU_verdef   SHT_SUNW_verdef
#define SHT_GNU_verneed  SHT_SUNW_verneed
#define SHT_GNU_versym   SHT_SUNW_versym

#define SHT_LOPROC       0x70000000    /* Processor-specific semantics, lo */
#define SHT_HIPROC       0x7FFFFFFF    /* Processor-specific semantics, hi */
#define SHT_LOUSER       0x80000000    /* Application-specific semantics */
/* #define SHT_HIUSER 0x8FFFFFFF    */ /* Application-specific semantics */
#define SHT_HIUSER 0xFFFFFFFF          /* New value, defined in Oct 4, 1999 Draft */

/* Values for section header, sh_flags field */

#define SHF_WRITE            (1 << 0) /* Writable data during execution */
#define SHF_ALLOC            (1 << 1) /* Occupies memory during execution */
#define SHF_EXECINSTR        (1 << 2) /* Executable machine instructions */
#define SHF_MERGE            (1 << 4) /* Data in this section can be merged */
#define SHF_STRINGS          (1 << 5) /* Contains null terminated character strings */
#define SHF_INFO_LINK        (1 << 6) /* sh_info holds section header table index */
#define SHF_LINK_ORDER       (1 << 7) /* Preserve section ordering when linking */
#define SHF_OS_NONCONFORMING (1 << 8) /* OS specifci processing required */


/* #define SHF_MASKOS 0x0F000000    */ /* OS-specific semantics */
#define SHF_MASKOS   0x0FF00000        /* New value, Oct 4, 1999 Draft */
#define SHF_MASKPROC 0xF0000000        /* Processor-specific semantics */

/* Special section indices, which may show up in st_shndx fields, among
   other places. */

#define SHN_UNDEF     0x0000 /* Undefined section reference */
#define SHN_LORESERVE 0xFF00 /* Begin range of reserved indices */
#define SHN_LOPROC    0xFF00 /* Begin range of appl-specific */
#define SHN_HIPROC    0xFF1F /* End range of appl-specific */
#define SHN_LOOS      0xFF20 /* OS specific semantics, lo */
#define SHN_HIOS      0xFF3F /* OS specific semantics, hi */
#define SHN_ABS       0xFFF1 /* Associated symbol is absolute */
#define SHN_COMMON    0xFFF2 /* Associated symbol is in common */
#define SHN_HIRESERVE 0xFFFF /* End range of reserved indices */


/* Values for program header, p_type field */

#define PT_NULL    0          /* Program header table entry unused */
#define PT_LOAD    1          /* Loadable program segment */
#define PT_DYNAMIC 2          /* Dynamic linking information */
#define PT_INTERP  3          /* Program interpreter */
#define PT_NOTE    4          /* Auxiliary information */
#define PT_SHLIB   5          /* Reserved, unspecified semantics */
#define PT_PHDR    6          /* Entry for header table itself */
#define PT_LOOS    0x60000000 /* OS-specific */
#define PT_HIOS    0x6fffffff /* OS-specific */
#define PT_LOPROC  0x70000000 /* Processor-specific */
#define PT_HIPROC  0x7FFFFFFF /* Processor-specific */


/* Program segment permissions, in program header p_flags field */

#define PF_X (1 << 0) /* Segment is executable */
#define PF_W (1 << 1) /* Segment is writable */
#define PF_R (1 << 2) /* Segment is readable */

/* #define PF_MASKOS  0x0F000000    */ /* OS-specific reserved bits */
#define PF_MASKOS   0x0FF00000         /* New value, Oct 4, 1999 Draft */
#define PF_MASKPROC 0xF0000000         /* Processor-specific reserved bits */


#define STB_LOCAL   0  /* Symbol not visible outside obj */
#define STB_GLOBAL  1  /* Symbol visible outside obj */
#define STB_WEAK    2  /* Like globals, lower precedence */
#define STB_LOOS    10 /* OS-specific semantics */
#define STB_HIOS    12 /* OS-specific semantics */
#define STB_LOPROC  13 /* Application-specific semantics */
#define STB_HIPROC  15 /* Application-specific semantics */

#define STT_NOTYPE  0  /* Symbol type is unspecified */
#define STT_OBJECT  1  /* Symbol is a data object */
#define STT_FUNC    2  /* Symbol is a code object */
#define STT_SECTION 3  /* Symbol associated with a section */
#define STT_FILE    4  /* Symbol gives a file name */
#define STT_COMMON  5  /* An uninitialised common block */
#define STT_LOOS    10 /* OS-specific semantics */
#define STT_HIOS    12 /* OS-specific semantics */
#define STT_LOPROC  13 /* Application-specific semantics */
#define STT_HIPROC  15 /* Application-specific semantics */


#define EI_NIDENT   16

#pragma pack(1)

typedef struct {
    char  e_ident[EI_NIDENT];
    WORD  e_type;
    WORD  e_machine;
    DWORD e_version;
    DWORD e_entry;
    DWORD e_phoff;
    DWORD e_shoff;
    DWORD e_flags;
    WORD  e_ehsize;
    WORD  e_phentsize;
    WORD  e_phnum;
    WORD  e_shentsize;
    WORD  e_shnum;
    WORD  e_shstrndx;
} Elf32_Ehdr;

typedef struct {
    DWORD sh_name;
    DWORD sh_type;
    DWORD sh_flags;
    DWORD sh_addr;
    DWORD sh_offset;
    DWORD sh_size;
    DWORD sh_link;
    DWORD sh_info;
    DWORD sh_addralign;
    DWORD sh_entsize;
} Elf32_Shdr;

typedef struct {
    DWORD p_type;
    DWORD p_offset;
    DWORD p_vaddr;
    DWORD p_paddr;
    DWORD p_filesz;
    DWORD p_memsz;
    DWORD p_flags;
    DWORD p_align;
} Elf32_Phdr;

typedef struct {
    DWORD st_name;
    DWORD st_value;
    DWORD st_size;
    BYTE  st_info;
    BYTE  st_other;
    WORD  st_shndx;
} Elf32_Sym;

#pragma pack()


struct Elf32_Info {
    FILE *      fh;      // File Handle
    Elf32_Ehdr  ehdr;    // ELF Header
    Elf32_Shdr *shdr;    // List of ELF Section Headers
    Elf32_Phdr *phdr;    // List of ELF Program Headers
    Elf32_Sym * sym;     // List of ELF Symbol Tables
    DWORD       sym_cnt; // Symbol Count
    char *      strtab;  // String Table
};

extern Elf32_Info Elf; // ELF Information

extern void ElfInit(void);   // ELF Initialization
extern void ElfUnInit(void); // ELF UnInitialization
extern BOOL ReadElf(void);   // Read ELF Information


#endif
